ai 사용 내역 : https://claude.ai/share/e855f2d7-8741-4279-8012-d0e35d966a62

In [14]:
import numpy as np
import pandas as pd

works = pd.DataFrame({
    '작가': ['김유정', '김유정', '김유정',
             '현진건', '현진건',
             '이상', '이상',
             '나도향', '나도향',
             '채만식', '이효석'],
    '작품': ['봄봄', '동백꽃', '만무방',
             '운수 좋은 날', 'B사감과 러브레터',
             '날개', '권태',
             '벙어리 삼룡이', '물레방아',
             '레디메이드 인생', '메밀꽃 필 무렵'],
    '연도': [1935, 1936, 1934,
             1924, 1925,
             1936, 1937,
             1925, 1925,
             1934, np.nan],   # 이효석 '메밀꽃 필 무렵' 발표연도가 누락됨
    '글자수': [12500, 9800, 13600,
              11200, 7400,
              18300, 14100,
              16800, 14500,
              22000, 11800],
})

authors = pd.DataFrame({
    '작가': ['김유정', '현진건', '이상', '나도향', '채만식', '이효석', '염상섭'],
    '생년': [1908, 1900, 1910, 1902, 1902, 1907, 1897],
    '몰년': [1937, 1943, 1937, 1926, 1950, 1942, 1963],
})

Q1

(a)

(1) 코드

In [15]:
print(works.shape)
print(works.dtypes)
print(works.describe())

(11, 4)
작가         str
작품         str
연도     float64
글자수      int64
dtype: object
                연도           글자수
count    10.000000     11.000000
mean   1931.100000  13818.181818
std       5.546771   4080.641661
min    1924.000000   7400.000000
25%    1925.000000  11500.000000
50%    1934.000000  13600.000000
75%    1935.750000  15650.000000
max    1937.000000  22000.000000


(2) 코드 실행 결과 : (11, 4)
작가         str
작품         str
연도     float64
글자수      int64
dtype: object, (11, 4)
작가         str
작품         str
연도     float64
글자수      int64
dtype: object
                연도           글자수
count    10.000000     11.000000
mean   1931.100000  13818.181818
std       5.546771   4080.641661
min    1924.000000   7400.000000
25%    1925.000000  11500.000000
50%    1934.000000  13600.000000
75%    1935.750000  15650.000000
max    1937.000000  22000.000000

(3) 설명 : 11행 4열이므로 shape 출력값이 (11,4)이며, 작가와 작품은 모두 문자열이라 str이다. (짧은 관찰) : NaN은 부동소수점이므로, 연도는 NaN이 섞여 float가 되고, 글자 수는 NaN이 없어 int로 유지된다. 

(b)

(1) 코드

In [16]:
works.isna().sum()

작가     0
작품     0
연도     1
글자수    0
dtype: int64

(2) 코드 실행 결과 : 작가     0
작품     0
연도     1
글자수    0
dtype: int64

(3) 설명 : 연도 열에만 1건의 결측치가 있다.

(c)

(1) 코드

In [17]:
works2 = works.fillna({'연도': 1936})
works2['연도'] = works2['연도'].astype(int)

print(works2.dtypes)
print(works2.tail(3))

작가       str
작품       str
연도     int64
글자수    int64
dtype: object
     작가        작품    연도    글자수
8   나도향      물레방아  1925  14500
9   채만식  레디메이드 인생  1934  22000
10  이효석  메밀꽃 필 무렵  1936  11800


(2) 코드 실행 결과 : 작가       str
작품       str
연도     int64
글자수    int64
dtype: object
     작가        작품    연도    글자수
8   나도향      물레방아  1925  14500
9   채만식  레디메이드 인생  1934  22000
10  이효석  메밀꽃 필 무렵  1936  11800

(3) 설명 : fillna(1936)으로 누락된 발표연도를 채우자 NaN이 사라졌고, astype(int)으로 연도의 dtype이 float64에서 int64로 바뀌었다. tail(3)에서 '메밀꽃 필 무렵'의 연도가 1936으로 채워진 것을 확인할 수 있다.

Q2

(a)

(1) 코드

In [18]:
after_1930 = works2[works2['연도'] >= 1930]
print(after_1930)

kim_or_lee = works2[works2['작가'].isin(['김유정', '이상'])]
print(kim_or_lee)

     작가        작품    연도    글자수
0   김유정        봄봄  1935  12500
1   김유정       동백꽃  1936   9800
2   김유정       만무방  1934  13600
5    이상        날개  1936  18300
6    이상        권태  1937  14100
9   채만식  레디메이드 인생  1934  22000
10  이효석  메밀꽃 필 무렵  1936  11800
    작가   작품    연도    글자수
0  김유정   봄봄  1935  12500
1  김유정  동백꽃  1936   9800
2  김유정  만무방  1934  13600
5   이상   날개  1936  18300
6   이상   권태  1937  14100


(2) 코드 실행 결과 : 작가        작품    연도    글자수
0   김유정        봄봄  1935  12500
1   김유정       동백꽃  1936   9800
2   김유정       만무방  1934  13600
5    이상        날개  1936  18300
6    이상        권태  1937  14100
9   채만식  레디메이드 인생  1934  22000
10  이효석  메밀꽃 필 무렵  1936  11800
    작가   작품    연도    글자수
0  김유정   봄봄  1935  12500
1  김유정  동백꽃  1936   9800
2  김유정  만무방  1934  13600
5   이상   날개  1936  18300
6   이상   권태  1937  14100

(3) 설명 : [works2['연도'] >= 1930]은 원소별 비교로 불 Series를 만들고, 이를 대괄호에 넣어 True인 행만 골라내는 불 인덱싱이다. isin(['김유정', '이상'])은 두 작가 중 하나(또는)에 해당하는 행을 선택한다. 

(b)

(1) 코드

In [19]:
def categorize(n: int) -> str:
    if n < 10000:
        return '짧음'
    elif n <= 15000:
        return '보통'
    else:
        return '긴'
    
works2['분량'] = works2['글자수'].apply(categorize)
works2[['작품', '글자수', '분량']]
    

,작품,글자수,분량
0,봄봄,12500,보통
1,동백꽃,9800,짧음
2,만무방,13600,보통
3,운수 좋은 날,11200,보통
4,B사감과 러브레터,7400,짧음
5,날개,18300,긴
6,권태,14100,보통
7,벙어리 삼룡이,16800,긴
8,물레방아,14500,보통
9,레디메이드 인생,22000,긴


(2) 코드 실행 결과 : <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>작품</th>
      <th>글자수</th>
      <th>분량</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>봄봄</td>
      <td>12500</td>
      <td>보통</td>
    </tr>
    <tr>
      <th>1</th>
      <td>동백꽃</td>
      <td>9800</td>
      <td>짧음</td>
    </tr>
    <tr>
      <th>2</th>
      <td>만무방</td>
      <td>13600</td>
      <td>보통</td>
    </tr>
    <tr>
      <th>3</th>
      <td>운수 좋은 날</td>
      <td>11200</td>
      <td>보통</td>
    </tr>
    <tr>
      <th>4</th>
      <td>B사감과 러브레터</td>
      <td>7400</td>
      <td>짧음</td>
    </tr>
    <tr>
      <th>5</th>
      <td>날개</td>
      <td>18300</td>
      <td>긴</td>
    </tr>
    <tr>
      <th>6</th>
      <td>권태</td>
      <td>14100</td>
      <td>보통</td>
    </tr>
    <tr>
      <th>7</th>
      <td>벙어리 삼룡이</td>
      <td>16800</td>
      <td>긴</td>
    </tr>
    <tr>
      <th>8</th>
      <td>물레방아</td>
      <td>14500</td>
      <td>보통</td>
    </tr>
    <tr>
      <th>9</th>
      <td>레디메이드 인생</td>
      <td>22000</td>
      <td>긴</td>
    </tr>
    <tr>
      <th>10</th>
      <td>메밀꽃 필 무렵</td>
      <td>11800</td>
      <td>보통</td>
    </tr>
  </tbody>
</table>
</div>

(3) 설명 : 벡터화로 안 되는 변환은 apply로 처리할 수 있다. Series.apply(categorize)로 원소마다 categorize 함수를 호출해 새로운 분량 열을 만들었다.

(c)

(1) 코드

In [20]:
works2.sort_values(by=['연도', '글자수'], ascending=[True, False])

,작가,작품,연도,글자수,분량
3,현진건,운수 좋은 날,1924,11200,보통
7,나도향,벙어리 삼룡이,1925,16800,긴
8,나도향,물레방아,1925,14500,보통
4,현진건,B사감과 러브레터,1925,7400,짧음
9,채만식,레디메이드 인생,1934,22000,긴
2,김유정,만무방,1934,13600,보통
0,김유정,봄봄,1935,12500,보통
5,이상,날개,1936,18300,긴
10,이효석,메밀꽃 필 무렵,1936,11800,보통
1,김유정,동백꽃,1936,9800,짧음


(2) 코드 실행 결과 : <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>작가</th>
      <th>작품</th>
      <th>연도</th>
      <th>글자수</th>
      <th>분량</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>3</th>
      <td>현진건</td>
      <td>운수 좋은 날</td>
      <td>1924</td>
      <td>11200</td>
      <td>보통</td>
    </tr>
    <tr>
      <th>7</th>
      <td>나도향</td>
      <td>벙어리 삼룡이</td>
      <td>1925</td>
      <td>16800</td>
      <td>긴</td>
    </tr>
    <tr>
      <th>8</th>
      <td>나도향</td>
      <td>물레방아</td>
      <td>1925</td>
      <td>14500</td>
      <td>보통</td>
    </tr>
    <tr>
      <th>4</th>
      <td>현진건</td>
      <td>B사감과 러브레터</td>
      <td>1925</td>
      <td>7400</td>
      <td>짧음</td>
    </tr>
    <tr>
      <th>9</th>
      <td>채만식</td>
      <td>레디메이드 인생</td>
      <td>1934</td>
      <td>22000</td>
      <td>긴</td>
    </tr>
    <tr>
      <th>2</th>
      <td>김유정</td>
      <td>만무방</td>
      <td>1934</td>
      <td>13600</td>
      <td>보통</td>
    </tr>
    <tr>
      <th>0</th>
      <td>김유정</td>
      <td>봄봄</td>
      <td>1935</td>
      <td>12500</td>
      <td>보통</td>
    </tr>
    <tr>
      <th>5</th>
      <td>이상</td>
      <td>날개</td>
      <td>1936</td>
      <td>18300</td>
      <td>긴</td>
    </tr>
    <tr>
      <th>10</th>
      <td>이효석</td>
      <td>메밀꽃 필 무렵</td>
      <td>1936</td>
      <td>11800</td>
      <td>보통</td>
    </tr>
    <tr>
      <th>1</th>
      <td>김유정</td>
      <td>동백꽃</td>
      <td>1936</td>
      <td>9800</td>
      <td>짧음</td>
    </tr>
    <tr>
      <th>6</th>
      <td>이상</td>
      <td>권태</td>
      <td>1937</td>
      <td>14100</td>
      <td>보통</td>
    </tr>
  </tbody>
</table>
</div>

(3) 설명 : ascending=[True, False]는 by 리스트와 순서가 대응하여 연도 오름차순, 같은 연도 안에서는 글자수 내림차순으로 정렬한다. 

Q3

(a)

(1) 코드

In [21]:
works2.groupby('작가')['글자수'].agg(['mean', 'count', 'sum'])

,mean,count,sum
작가,,,
김유정,11966.666667,3,35900
나도향,15650.000000,2,31300
이상,16200.000000,2,32400
이효석,11800.000000,1,11800
채만식,22000.000000,1,22000
현진건,9300.000000,2,18600


(2) 코드 실행 결과 : <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>mean</th>
      <th>count</th>
      <th>sum</th>
    </tr>
    <tr>
      <th>작가</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>김유정</th>
      <td>11966.666667</td>
      <td>3</td>
      <td>35900</td>
    </tr>
    <tr>
      <th>나도향</th>
      <td>15650.000000</td>
      <td>2</td>
      <td>31300</td>
    </tr>
    <tr>
      <th>이상</th>
      <td>16200.000000</td>
      <td>2</td>
      <td>32400</td>
    </tr>
    <tr>
      <th>이효석</th>
      <td>11800.000000</td>
      <td>1</td>
      <td>11800</td>
    </tr>
    <tr>
      <th>채만식</th>
      <td>22000.000000</td>
      <td>1</td>
      <td>22000</td>
    </tr>
    <tr>
      <th>현진건</th>
      <td>9300.000000</td>
      <td>2</td>
      <td>18600</td>
    </tr>
  </tbody>
</table>
</div>

(3) 설명 : 작가로 행들을 묶고(Split) 각 그룹에 같은 함수 mean, count, sum을 적용해(apply) 결과를 하나의 표로 합쳤다(Combine). 

(b)

(1) 코드

In [22]:
works2['분량'].value_counts()

분량
보통    6
긴     3
짧음    2
Name: count, dtype: int64

(2) 코드 실행 결과 : 분량
보통    6
긴     3
짧음    2
Name: count, dtype: int64

(3) 설명 : value_counts는 분량 열의 빈도를 큰 순으로 돌려준다. 

(c)

(1) 코드

In [23]:
works2.groupby(['작가', '분량']).size()

작가   분량
김유정  보통    2
     짧음    1
나도향  긴     1
     보통    1
이상   긴     1
     보통    1
이효석  보통    1
채만식  긴     1
현진건  보통    1
     짧음    1
dtype: int64

(2) 코드 실행 결과 : 작가   분량
김유정  보통    2
     짧음    1
나도향  긴     1
     보통    1
이상   긴     1
     보통    1
이효석  보통    1
채만식  긴     1
현진건  보통    1
     짧음    1
dtype: int64

(3) 설명 : 그룹 키를 리스트로 주면 다층 인덱스가 된다. (관찰) : 가장 많은 작품 수를 가진 (작가, 분량) 조합은 (김유정, 보통)으로 2건이며, 이는 김유정이 10,000 이상 15,000 이하의 '보통' 분량의 작품을 주로 썼음을 시사한다. 

Q4

(a)

(1) 코드

In [24]:
full = pd.merge(works2, authors, on='작가', how='left')

print(full)
full.shape

     작가         작품    연도    글자수  분량    생년    몰년
0   김유정         봄봄  1935  12500  보통  1908  1937
1   김유정        동백꽃  1936   9800  짧음  1908  1937
2   김유정        만무방  1934  13600  보통  1908  1937
3   현진건    운수 좋은 날  1924  11200  보통  1900  1943
4   현진건  B사감과 러브레터  1925   7400  짧음  1900  1943
5    이상         날개  1936  18300   긴  1910  1937
6    이상         권태  1937  14100  보통  1910  1937
7   나도향    벙어리 삼룡이  1925  16800   긴  1902  1926
8   나도향       물레방아  1925  14500  보통  1902  1926
9   채만식   레디메이드 인생  1934  22000   긴  1902  1950
10  이효석   메밀꽃 필 무렵  1936  11800  보통  1907  1942


(11, 7)

(2) 코드 실행 결과 : 작가         작품    연도    글자수  분량    생년    몰년
0   김유정         봄봄  1935  12500  보통  1908  1937
1   김유정        동백꽃  1936   9800  짧음  1908  1937
2   김유정        만무방  1934  13600  보통  1908  1937
3   현진건    운수 좋은 날  1924  11200  보통  1900  1943
4   현진건  B사감과 러브레터  1925   7400  짧음  1900  1943
5    이상         날개  1936  18300   긴  1910  1937
6    이상         권태  1937  14100  보통  1910  1937
7   나도향    벙어리 삼룡이  1925  16800   긴  1902  1926
8   나도향       물레방아  1925  14500  보통  1902  1926
9   채만식   레디메이드 인생  1934  22000   긴  1902  1950
10  이효석   메밀꽃 필 무렵  1936  11800  보통  1907  1942 (11, 7)

(3) 설명 : (확인 질문) : authors에는 있지만 works2에는 없는 작가 염상섭은 full에 포함되지 않는다. how='left'는 works2를 기준으로 모든 행을 살리고 authors에서는 짝이 맞는 행만 가져오므로 authors에만 있는 행은 말없이 사라진다.

(b)

(1) 코드

In [25]:
full['집필 당시 나이'] = full['연도'] - full['생년']

full[['작가', '작품', '연도', '생년', '집필 당시 나이']]

,작가,작품,연도,생년,집필 당시 나이
0,김유정,봄봄,1935,1908,27
1,김유정,동백꽃,1936,1908,28
2,김유정,만무방,1934,1908,26
3,현진건,운수 좋은 날,1924,1900,24
4,현진건,B사감과 러브레터,1925,1900,25
5,이상,날개,1936,1910,26
6,이상,권태,1937,1910,27
7,나도향,벙어리 삼룡이,1925,1902,23
8,나도향,물레방아,1925,1902,23
9,채만식,레디메이드 인생,1934,1902,32


(2) 코드 실행 결과 : <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>작가</th>
      <th>작품</th>
      <th>연도</th>
      <th>생년</th>
      <th>집필 당시 나이</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>김유정</td>
      <td>봄봄</td>
      <td>1935</td>
      <td>1908</td>
      <td>27</td>
    </tr>
    <tr>
      <th>1</th>
      <td>김유정</td>
      <td>동백꽃</td>
      <td>1936</td>
      <td>1908</td>
      <td>28</td>
    </tr>
    <tr>
      <th>2</th>
      <td>김유정</td>
      <td>만무방</td>
      <td>1934</td>
      <td>1908</td>
      <td>26</td>
    </tr>
    <tr>
      <th>3</th>
      <td>현진건</td>
      <td>운수 좋은 날</td>
      <td>1924</td>
      <td>1900</td>
      <td>24</td>
    </tr>
    <tr>
      <th>4</th>
      <td>현진건</td>
      <td>B사감과 러브레터</td>
      <td>1925</td>
      <td>1900</td>
      <td>25</td>
    </tr>
    <tr>
      <th>5</th>
      <td>이상</td>
      <td>날개</td>
      <td>1936</td>
      <td>1910</td>
      <td>26</td>
    </tr>
    <tr>
      <th>6</th>
      <td>이상</td>
      <td>권태</td>
      <td>1937</td>
      <td>1910</td>
      <td>27</td>
    </tr>
    <tr>
      <th>7</th>
      <td>나도향</td>
      <td>벙어리 삼룡이</td>
      <td>1925</td>
      <td>1902</td>
      <td>23</td>
    </tr>
    <tr>
      <th>8</th>
      <td>나도향</td>
      <td>물레방아</td>
      <td>1925</td>
      <td>1902</td>
      <td>23</td>
    </tr>
    <tr>
      <th>9</th>
      <td>채만식</td>
      <td>레디메이드 인생</td>
      <td>1934</td>
      <td>1902</td>
      <td>32</td>
    </tr>
    <tr>
      <th>10</th>
      <td>이효석</td>
      <td>메밀꽃 필 무렵</td>
      <td>1936</td>
      <td>1907</td>
      <td>29</td>
    </tr>
  </tbody>
</table>
</div>

(3) 설명 : full['연도'] - full['생년']은 두 열을 원소별로 빼는 벡터화 연산이다. 한 줄로 모든 원소에 한꺼번에 적용되므로 파이썬 for로 한 칸씩 도는 것보다 훨씬 빠르고 의도가 명확하다. pandas에서도 numpy처럼 벡터화 연산을 할 수 있다.

(c)

(1) 코드

In [26]:
full.groupby('작가')['집필 당시 나이'].mean().sort_values()

작가
나도향    23.0
현진건    24.5
이상     26.5
김유정    27.0
이효석    29.0
채만식    32.0
Name: 집필 당시 나이, dtype: float64

(2) 코드 실행 결과 : 작가
나도향    23.0
현진건    24.5
이상     26.5
김유정    27.0
이효석    29.0
채만식    32.0
Name: 집필 당시 나이, dtype: float64

(3) 설명 : groupby 결과에 sort_values를 이어 붙여 오름차순 정렬했다. (해석) : 가장 어린 나이에 작품을 발표한 작가는 나도향이며, 평균 23살에 작품을 썼다. 가장 늦은 나이에 발표한 작가는 채만식이다.